### PR TITLE
Leap forward to 7.4 with preparing support for 8.X

### PR DESCRIPTION
**Pull request by HaiTo**
# Why
We are thinking of adding support for PHP 8.0.
However, it should require a major change.
So, we'll first drop off Support for 5.X and Support up to 7.4.

# What did 
To make it as easy as possible for further development.
**Also, I break backward compatibility.**

1. Drop EOL PHP supports.
2. Support 7.3, 7.4 PHP features.
3. "type declarations/hints" wherever possible.
4. Upgrade PHP Unit to 8.X
    * With fixing XML.
5. Bump up All dependencies

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -3,10 +3,9 @@ language: php
 sudo: false
 
 php:
-  - 5.6
-  - 7.0
-  - 7.1
-  - 7.2
+  - 7.3
```
Comment (HaiTo) [inline .travis.yml:6]: Ref: https://www.php.net/supported-versions.php

File: .travis.yml [hunk 2]
```diff
@@ -18,4 +17,6 @@ script:
 
 matrix:
   allow_failures:
-    - php: hhvm
+    - php:
+      - hhvm
+      - 8.0
```
Comment (HaiTo) [inline .travis.yml:22]: The current PR doesn't require for 8.0

File: composer.json [hunk 3]
```diff
@@ -4,15 +4,16 @@
...
+        "pimple/pimple": "3.3.x",
+        "psr/log": "1.1.*",
+        "monolog/monolog": "2.2.x",
+        "symfony/event-dispatcher": "5.2.x",
+        "php": ">=7.3"
     },
     "require-dev": {
-        "phpunit/phpunit": "^5.7 || ^6.4 || ^7.4",
-        "friendsofphp/php-cs-fixer": "^2.7 !=2.16.1"
+        "phpunit/phpunit": "8.*",
```
Comment (HaiTo) [inline composer.json:15]: 7.3 doesn't support to 9.* yet

File: composer.json [hunk 4]
```diff
@@ -4,15 +4,16 @@
...
-        "pimple/pimple": "3.0.x",
-        "psr/log": "1.0.0",
-        "monolog/monolog": "1.17.x",
-        "symfony/event-dispatcher": "3.0.x"
+        "twig/twig": "3.3.x",
+        "pimple/pimple": "3.3.x",
+        "psr/log": "1.1.*",
+        "monolog/monolog": "2.2.x",
+        "symfony/event-dispatcher": "5.2.x",
+        "php": ">=7.3"
```
Comment (HaiTo) [inline composer.json:12]: Add PHP version expectation

File: phpunit.xml.dist [hunk 5]
```diff
@@ -1,26 +1,50 @@
...
-    </filter>
     <logging>
+        <!-- Remove <log> elements when PHP unit bumped 9.X -->
         <log type="junit" target="build/junit/junit.xml" />
-        <log type="coverage-clover" target="build/coverage/clover.xml"/>
-        <log type="coverage-html" target="build/coverage/html/"/>
-        <log type="coverage-text" target="php://stdout" showUncoveredFiles="true"/>
+        <log type="coverage-html" target="build/coverage/html/" />
+        <log type="coverage-clover" target="build/coverage/clover.xml" />
+        <!-- For PHP unit 9.X --> <!--
```
Comment (HaiTo) [inline phpunit.xml.dist:16]: Hints for 9.X

File: src/Components/ArrayResourceTrait.php [hunk 6]
```diff
@@ -10,21 +7,21 @@ trait ArrayResourceTrait
...
+     * @param mixed $key
+     * @param mixed $value
      */
-    public function setResource($key, $value)
+    public function setResource($key, $value): void
     {
         $key_parts = explode('.', $key);
         $ref_value = &$this->_array_resource;
-        foreach ($key_parts as $key) {
+        foreach ($key_parts as $key_part) {
```
Comment (HaiTo) [inline src/Components/ArrayResourceTrait.php:17]: Solve deprecated for arguments from function.

File: src/Dispatcher.php [hunk 7]
```diff
@@ -175,23 +168,19 @@ public function handleError(\Exception $errors)
             $response = $this->container['response'];
         }
 
-        $action_result = "";
```
Comment (HaiTo) [inline src/Dispatcher.php:178]: the `$action_result` is always defined during if statements.

File: src/Dispatcher.php [hunk 8]
```diff
@@ -278,11 +270,11 @@ protected function dispatchRouter($method, $path)
         case RouteDispatcher::NOT_FOUND:
             $logger->debug('Routing failed. Not Found.');
             throw new HttpNotFoundException('404 Not Found');
-            break;
```
Comment (HaiTo) [inline src/Dispatcher.php:281]: Already throw-ed

File: src/Dispatcher.php [hunk 9]
```diff
@@ -339,17 +331,27 @@ protected function finishRequest(Response $response)
...
+     * @param string $env
+     * @return array|false|mixed|string
+     */
+    public static function getEnv(string $env = 'production')
     {
         if (isset($_SERVER['DIET_ENV'])) {
-            $env = $_SERVER['DIET_ENV'];
-        } elseif (getenv('DIET_ENV')) {
-            $env = getenv('DIET_ENV');
+            return $_SERVER['DIET_ENV'];
```
Comment (HaiTo) [inline src/Dispatcher.php:341]: Be guard

File: src/Events/DietcubeEventAbstract.php [hunk 10]
```diff
@@ -1,14 +1,11 @@
...
- *
- */
 
 namespace Dietcube\Events;
 
 use Dietcube\Application;
 use Symfony\Component\EventDispatcher\Event;
 
-class DietcubeEventAbstract extends Event
+abstract class DietcubeEventAbstract extends Event
```
Comment (HaiTo) [inline src/Events/DietcubeEventAbstract.php:8]: It's expected an abstract class. 
It's one of the big breaking changes from the current DietCube architecture.

File: src/Router.php [hunk 11]
```diff
@@ -124,28 +128,34 @@ public function url($handler, array $data = [], array $query_params = [], $is_ab
     }
 
     /**
-     * @return $this
```
Comment (HaiTo) [inline src/Router.php:127]: It's not $this

File: src/Twig/DietcubeExtension.php [hunk 12]
```diff
@@ -34,7 +36,7 @@ public function getFunctions()
      * @param bool $is_absolute
      * @return string url
      */
-    public function url($handler, array $data = [], array $query_params = [], $is_absolute = false)
+    public function url(string $handler, array $data = [], array $query_params = [], $is_absolute = false): string
```
Comment (HaiTo) [inline src/Twig/DietcubeExtension.php:39]: I trusted PHPDoc...

File: tests/RouterTest.php [hunk 13]
```diff
@@ -38,33 +38,30 @@ public function testDispatchPageNotFound()
         $this->assertSame([\FastRoute\Dispatcher::NOT_FOUND], $route_info);
     }
 
-    /**
-     * @expectedException \RuntimeException
```
Comment (HaiTo) [inline tests/RouterTest.php:42]: `@expectedException` is already removed at the new PHPUnit.

**Review by HaiTo — COMMENTED** [review]
Review hints

File: .travis.yml [hunk 14]
```diff
@@ -18,4 +17,6 @@ script:
 
 matrix:
   allow_failures:
-    - php: hhvm
+    - php:
+      - hhvm
```
Comment (DQNEO) [inline .travis.yml:21]: Same as above

File: .travis.yml [hunk 15]
```diff
@@ -3,10 +3,9 @@ language: php
...
 
 php:
-  - 5.6
-  - 7.0
-  - 7.1
-  - 7.2
+  - 7.3
+  - 7.4
+  - 8.0
   - hhvm
```
Comment (DQNEO) [inline .travis.yml:9]: hhvm is not compatible to PHP any more.
We should remove hhvm.
https://hhvm.com/blog/2019/02/11/hhvm-4.0.0.html

File: composer.json (hunk 3 above)
Comment (DQNEO) [reply composer.json:15]: Defining both would be better ? e.g. `"^8.0 || ^9.0"`
Then the composer will automatically select the suitable one.

File: src/Application.php [hunk 16]
```diff
@@ -47,15 +65,12 @@ public function __construct($app_root, $env)
         $this->dirs = $this->getDefaultDirs();
     }
 
-    /**
-     * @return Container
-     */
-    public function getContainer()
+    public function getContainer(): ?Container
```
Comment (DQNEO) [inline src/Application.php:68]: Why nullable ?
The original `@return Container` indicates this should not be null.

File: src/Application.php [hunk 17]
```diff
@@ -111,96 +128,111 @@ public function getRoute()
         return new $route_class;
     }
 
-    /**
-     * @return Config
-     */
-    public function getConfig()
+    public function getConfig(): ?Config
```
Comment (DQNEO) [inline src/Application.php:131]: wny nullable ?

File: src/Application.php [hunk 18]
```diff
@@ -111,96 +128,111 @@ public function getRoute()
...
     }
 
-    public function setDir($dirname, $path)
+    public function setDir($dirname, $path): void
     {
         $this->dirs[$dirname] = $path;
     }
 
+    /**
+     * @return mixed
```
Comment (DQNEO) [inline src/Application.php:142]: I don't this this `mixed` is useful.
It's not clear whether you want to keep it "mixed" permanently, or  just you want to explain "what is the current behavior".
As for the latter, "mixed" is obvious.

File: composer.json (hunk 3 above)
Comment (sanmai) [reply composer.json:15]: 9.x works on 7.3, so it'll never choose 8.x

**Comment by HaiTo** [comment]
Close at here

**Pull request closed without merging**